Update Contribution Docs, and Add Codespaces Docs (#1336)
demisto/content-docs: docs/contributing/contributing.md

1. Begin by designing your contribution. We recommend to follow the [design guidelines](../concepts/design) to identify what you want to build and make sure it is aligned with our best practices. Also check out the [design tutorial](../tutorials/tut-design).
2. Make sure you have all the [Contributing Requirements](../contributing/contrib-requirements) satisfied.
3. Setup a development environment by following the short [Development Setup](../concepts/dev-setup) guide, the more detailed [tutorial](../tutorials/tut-setup-dev), 
    or [use a pre-configured GitHub Codespace environment](../tutorials/tut-setup-dev-codespace.md).
4. Follow the [Content Pack Structure](../packs/packs-format) to build your contribution.  
    The "[demisto-sdk init](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/init/README.md)" command will help you create the basic structure.
5. If you are updating an **existing** content pack, make you have the latest available version installed before proceeding.

> 6. Depending on the content entities you need to build, navigate to the specific section of this website for details.  

Author: I think it's referring to the documentation site (xsoar.pan.dev) the article will be hosted on, but it definitely can be worded better.
I'll update that.

:::

:::note
If you are an XSOAR Technology Partner, make the [Cortex XSOAR Alliances Team](mailto:soar.alliances@paloaltonetworks.com) has reviewed your planned contribution, and that you have a *Partner ID* to associate your Pull Request to.
:::

## Pull Request Submission Process
### Prerequisites
Before submitting a Pull Request to the [Cortex XSOAR GitHub Repository](https://github.com/demisto/content), the following requirements must be met:
1. Make sure to check the content contribution [checklist](../contributing/checklist), and assure that all the required files exist and are properly formatted.
2. Create and upload your contribution to a GitHub **fork** of the XSOAR Content repository, and submit your changes **on a new branch** (do **not** push your changes to the *master* branch).
3. Validate your content using [demisto-sdk validate](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/validate/README.md):  
    `demisto-sdk validate -i Packs/YourPackName`

> 4. If your contribution has Integrations or Scripts, run linting (which also runs [unit-tests](../tutorials/tut-setup-dev#step-5-run-the-linter-and-unit-tests)) by running:  

Reviewer: ```suggestion
4. If your contribution has integrations or scripts, run linting (which also runs [unit-tests](../tutorials/tut-setup-dev#step-5-run-the-linter-and-unit-tests)) by running:  
```

    `demisto-sdk lint -i Packs/YourPakName`
5. Create a short demo video presenting the product and your pack, and link it.  

>     This will help our reviewers to better understand what the product is used for, and the newly contributed content integrates with it.

Reviewer: ```suggestion
    This will help our reviewers to better understand what the product is used for, and how the newly contributed content integrates with it.
```

![Creating a fork](../doc_imgs/contributing/create-a-new-fork.gif)

### Pull Request Creation

> After all the prerequisites are met, and you are ready to create your pull request, commit and push your work to the branch you have created in your forked repository (if you haven't already done so).  

Reviewer: ```suggestion
After all the prerequisites are met, and you are ready to create your pull request, commit and push your work to the branch you created in your forked repository (if you haven't already done so).  
```

> Then on the main page of your forked repository, click on the **Compare & pull request** button to open a new Pull Request (on the original Cortex XSOAR Content repository).  

Reviewer: ```suggestion
Then on the main page of your forked repository, click **Compare & pull request** to open a new Pull Request (on the original Cortex XSOAR Content repository).  
```

> Fill the description template, and click on **Create pull request**.

Reviewer: ```suggestion
Fill in the description template, and click **Create pull request**.
```

For additional information on how to create a Pull Request from a fork, refer to [official GitHub documentation](https://help.github.com/articles/creating-a-pull-request-from-a-fork).

### Post-Submission
After submitting your pull request, sign our [Contributor License Agreement (CLA)](https://github.com/demisto/content/blob/master/docs/cla.pdf), and monitor your pull request.  

Our Content team will review your code, and might ask questions and request changes.  

In order to establish a proper release timeframe for your contribution, you are required to respond (and apply the requested changes, if any) within 14 days.  
Stale pull requests will be closed.

At some point, when the review process is complete (or close to completion), we'll ask to schedule a meeting for an interactive demo.  

Prepare for it, and make sure you have a working instance of Cortex XSOAR, with your contributed pack fully configured.  
Check out our [Contribution Demo Page](../contributing/demo-prep) for additional information.

> :::caution

Reviewer: Caution is not the correct word to use.  It insinuate that something bad will happen.
Use "Note:"  or "Important:"

As part of the pull request template, you will be asked to fill in the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA).  
Make sure to do so, as without it, we will not be able to start the review process for your contribution.
:::

:::info
If after signing the CLA it still shows as unsigned, refer to the [FAQ](../concepts/faq#cla-is-pending-even-though-i-signed-the-agreement) for possible solutions.
:::

> For more details on how properly create and manage your pull request, check out the dedicated [Pull Request Conventions](../contributing/conventions) article.

Reviewer: ```suggestion
For more details on how to properly create and manage your pull request, check out the dedicated [Pull Request Conventions](../contributing/conventions) article.
```

---
id: contributing
title: Contributing XSOAR Content
---

This article describes the contribution process for Cortex XSOAR content.  

> Contributing allows clients / partners who made custom content, enhancements to existing content, and bug fixes, to share them with the community by making them publicly available in [Marketplace](/marketplace).  

Author: ```suggestion
Contributing allows clients / partners who made custom content, enhancements to existing content, and bug fixes, to share them with the community by making them publicly available in the [Cortex XSOAR Marketplace](https://cortex.marketplace.pan.dev/marketplace/).  
```

Content can be either partner or community supported, and can be either [free or paid](../partners/premium-packs#pricing).

All content (except for paid content packs) is open source, and is hosted on the [Cortex XSOAR Content GitHub Repository](https://github.com/demisto/content), licensed under the [MIT license](https://github.com/demisto/content/blob/master/LICENSE).  

**If you have any questions or need support, feel free to reach out to us on the `#demisto-developers` channel on our [Slack DFIR Community](https://www.demisto.com/community/).**  
:::note
This article is only for content contributions.
For documentation contributions, refer to [Documentation Contributions](docs-contrib).  
:::

## Content Support Types

> Contributions can be either *officially supported* (by Palo Alto Networks, a Technology Partner, or an individual), or *community supported*.  

Author: I got the idea from the following pre-existing sections (which was removed):

> Contributions can be either officially supported (by Palo Alto Networks, a Technology Partner, or an individual developer), or *community* supported: the former means that when a customer has a problem with the content, they will have an email address or web site to reach out to, and they will expect an answer. When the contribution is *community* supported, it's not required to provide support (customers will ask questions on our [Live Community Forum](https://live.paloaltonetworks.com/t5/cortex-xsoar-discussions/bd-p/Cortex_XSOAR_Discussions)) and the quality requirements for the contribution to be accepted will be lower.

But yeah, "individual" seems out-of place (what if it's a group of multiple people that aren't a parnter?).
Would:
```suggestion
Contributions can be either *officially supported* (by Palo Alto Networks, or a Technology Partner), or *community supported*.  
```
be accurate?

### Officially Supported Packs

> Officially supported packs include an email address or a website of the contributor for customers to contact for support (that they will be referred to in case of an issue), and have a more strict quality-control process.

Author: ```suggestion
Officially supported packs have a more strict quality-control process, and provide an email address or a website of their owner that customers can reach out to (and will be referred to) in case of an issue.
```

### Community Supported Packs
Community supported packs on the other hand, do not have a support contact (customers will be able to ask questions on our [Live Community Forum](https://live.paloaltonetworks.com/t5/cortex-xsoar-discussions/bd-p/Cortex_XSOAR_Discussions)), and will have a less strict quality-control process.

:::info
Contributed content for Palo Alto Networks products (e.g., PAN-OS, XDR, etc.) will be adopted and supported by the Cortex XSOAR Content team.  
:::

## Contribution Methods

> After creating new content to contribute, you must submit your content to Palo Alto Networks. The Cortex XSOAR Content Team will review and approve it before it becomes available to customers.

Author: ```suggestion
After creating new content, in order to contribute it, you will need to submit it for the Cortex XSOAR Content team to review. Once approved, it will become available for all customers to download.
```

Before starting to work on a contribution, it is recommended to go over the [Contribution SLA](../contributing/sla) article.

**There are several ways to submit contributions:**

### Cortex XSOAR Content GitHub Repository (Pull Request)
Contribute by creating a fork and opening a pull request on the [Cortex XSOAR Content GitHub repository](https://github.com/demisto/content).  
Use this method in the following scenarios:
* You are a Technology Partner contributing with *partner* supported new content.
* Your contribution is big and contains lots of different parts (integrations, scripts, playbooks, layouts, etc.) that is likely to lead to a complex review process.
* You are proficient with GitHub.

:::tip
For contributing through GitHub, we recommend to use a GitHub Codespace, which will provide you with a pre-configured ready-to-use development environment.  
This method is still experimental, but will make the contribution, development, and review processes much easier.  

For more information and a step-by-step guide, see the dedicated [GitHub Codespace Setup tutorial](../tutorials/tut-setup-dev-codespace.md).

> :::

Author: Wouldn't a link for the "[Set Up Your Dev Environment](https://xsoar-pan-dev--pull-request-1336-0i25g2kq.web.app/docs/tutorials/tut-setup-dev)" doc would be more fitting (instead of the first one)?

### Cortex XSOAR Marketplace

> Contribute from within Cortex XSOAR's UI using [Marketplace](../contributing/marketplace).  

Author: > Cortex XSOAR's UI using the XSOAR Marketplace.

I think specifying "XSOAR" again is redundant here.

```suggestion
Contribute from within Cortex XSOAR's UI using the Marketplace.
```

This method is more simple, and doesn't require you to set up a development environment or be familiar with Git.  

Contributing from the marketplace has several limitations:

> * The built-in editor is very limited and does not support features that full-fledged IDEs have.

Author: ```suggestion
* The built-in editor has limitations, and does not support some of the features that can be found in full-fledged IDE.
```

> * Documentation files (README, instance configuration help, etc.) cannot be created or edited.

Author: I think this is redundant, as it already says "Documentation files" before that, and the parentheses are just to give some examples for such files.

```suggestion
* Documentation files (README, integration's description, etc.) cannot be created or edited.
```

Then within a few minutes, the target branch will be replaced from `master` to a different custom one.
This is on purpose, and should not be changed.
:::

### Post-Submission
Before the review process begins, it is mandatory for you to complete two steps:

1. Sign our [Contributor License Agreement (CLA)](https://github.com/demisto/content/blob/master/docs/cla.pdf): Please read and sign our CLA.
2. Fill out the [registration form](https://forms.gle/XDfxU4E61ZwEESSMA).

Once you have created your pull request, an automated comment will be posted on your pull request with a link to the registration form, 
and a tag of a member of our contribution team who will review your code.  

Once these requirements are fulfilled, our assigned contributions team member will start reviewing your code.  

During or near the completion of the review process, we may request scheduling a meeting for an interactive demo.  
Prepare for it, and make sure you have a working instance of Cortex XSOAR, with your contributed pack fully configured and ready to be presented.  
Check out our [Contribution Demo Page](../contributing/demo-prep) for additional information about the demo process.

> :::caution

Reviewer: ```suggestion
:::info
```

In order to establish a proper release timeframe for your contribution, 
you are required to respond (and apply requested changes, if any) within 14 days.  

> Stale pull with no response within this timeframe will be closed.

Reviewer: ```suggestion
Stale pull requests with no response within this timeframe will be closed.
```
